File: alica_engine/src/engine/modelmanagement/factories/StateFactory.cpp
#include "engine/modelmanagement/factories/StateFactory.h"
#include "engine/model/State.h"
#include "engine/model/ConfAbstractPlanWrapper.h"
#include "engine/modelmanagement/Strings.h"
#include "engine/modelmanagement/factories/VariableBindingFactory.h"
#include "engine/modelmanagement/factories/ConfAbstractPlanWrapperFactory.h"

namespace alica
{
State* StateFactory::create(const YAML::Node& stateNode)
{
    State* state = new State();
    state->_type = State::StateType::NORMAL;
    Factory::setAttributes(stateNode, state);
    Factory::storeElement(state, alica::Strings::state);
    state->_inPlan = (Plan*) Factory::getElement(Factory::getReferencedId(stateNode[alica::Strings::parentPlan]));

    if (Factory::isValid(stateNode[alica::Strings::inTransitions])) {
        const YAML::Node& inTransitions = stateNode[alica::Strings::inTransitions];
        for (YAML::const_iterator it = inTransitions.begin(); it != inTransitions.end(); ++it) {
            Factory::stateInTransitionReferences.push_back(std::pair<int64_t, int64_t>(state->getId(), Factory::getReferencedId(*it)));
        }
    }
    if (Factory::isValid(stateNode[alica::Strings::outTransitions])) {
        const YAML::Node& outTransitions = stateNode[alica::Strings::outTransitions];
        for (YAML::const_iterator it = outTransitions.begin(); it != outTransitions.end(); ++it) {
            Factory::stateOutTransitionReferences.push_back(std::pair<int64_t, int64_t>(state->getId(), Factory::getReferencedId(*it)));
        }
    }
    if (Factory::isValid(stateNode[alica::Strings::confAbstractPlanWrappers])) {
        const YAML::Node& confAbstractPlanWrappers = stateNode[alica::Strings::confAbstractPlanWrappers];
        for (YAML::const_iterator it = confAbstractPlanWrappers.begin(); it != confAbstractPlanWrappers.end(); ++it) {
            state->_confAbstractPlanWrappers.push_back(ConfAbstractPlanWrapperFactory::create(*it));
        }
    }
    if (Factory::isValid(stateNode[alica::Strings::variableBindings])) {
        const YAML::Node& variableBindings = stateNode[alica::Strings::variableBindings];
        for (YAML::const_iterator it = variableBindings.begin(); it != variableBindings.end(); ++it) {
            state->_variableBindingGrp.push_back(VariableBindingFactory::create(*it));
        }
    }
    return state;
}

void StateFactory::attachReferences()
{
    VariableBindingFactory::attachReferences();
    ConfAbstractPlanWrapperFactory::attachReferences();

    // stateInTransitionReferences
    for (std::pair<int64_t, int64_t> pairs : Factory::stateInTransitionReferences) {
        Transition* t = (Transition*) Factory::getElement(pairs.second);
        State* st = (State*) Factory::getElement(pairs.first);
        st->_inTransitions.push_back(t);
    }
    Factory::stateInTransitionReferences.clear();

    // stateOutTransitionReferences
    for (std::pair<int64_t, int64_t> pairs : Factory::stateOutTransitionReferences) {
        State* st = (State*) Factory::getElement(pairs.first);
        Transition* t = (Transition*) Factory::getElement(pairs.second);
        st->_outTransitions.push_back(t);
    }
    Factory::stateOutTransitionReferences.clear();
}
} // namespace alica

File: alica-supplementary/plan-conversion-tool/src/main/java/com/rapyutarobotics/alica/factories/StateFactory.java
package com.rapyutarobotics.alica.factories;

import com.rapyutarobotics.alica.ConversionProcess;
import com.rapyutarobotics.alica.Tags;
import de.unikassel.vs.alica.planDesigner.alicamodel.*;
import javafx.util.Pair;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;

public class StateFactory extends Factory {

    public static State create(Element stateNode, ConversionProcess cp) {
        State state = new State();
        Factory.setAttributes(stateNode, state);
        cp.addElement(state);
        NodeList inTransitionNodes = stateNode.getElementsByTagName(Tags.INTRANSITIONS);
        for (int i = 0; i  < inTransitionNodes.getLength(); i++) {
            Element inTransitionNode = (Element) inTransitionNodes.item(i);
            cp.stateInTransitionReferences.put(state.getId(), cp.getReferencedId(inTransitionNode.getTextContent()));
        }
        NodeList outTransitionNodes = stateNode.getElementsByTagName(Tags.OUTTRANSITIONS);
        for (int i = 0; i  < outTransitionNodes.getLength(); i++) {
            Element outTransitionNode = (Element) outTransitionNodes.item(i);
            cp.stateOutTransitionReferences.put(state.getId(), cp.getReferencedId(outTransitionNode.getTextContent()));
        }
        NodeList abstractPlanNodes = stateNode.getElementsByTagName(Tags.PLANS);
        for (int i = 0; i  < abstractPlanNodes.getLength(); i++) {
            Element abstractPlanNode = (Element) abstractPlanNodes.item(i);
            state.addConfAbstractPlanWrapper(ConfAbstractPlanWrapperFactory.create(abstractPlanNode, cp));
        }
        NodeList variableBindingNodes = stateNode.getElementsByTagName(Tags.VARIABLEBINDING);
        for (int i = 0; i < variableBindingNodes.getLength(); i++) {
            Element variableBindingNode = (Element) variableBindingNodes.item(i);
            state.addVariableBinding(VariableBindingFactory.create(variableBindingNode, cp));
        }

        return state;
    }

    public static void attachReferences(ConversionProcess cp) {
        VariableBindingFactory.attachReferences(cp);
        ConfAbstractPlanWrapperFactory.attachReferences(cp);

        for (Pair<Long, Long> entry : cp.stateInTransitionReferences.getEntries()) {
            State state = (State) cp.getElement(entry.getKey());
            Transition transition = (Transition) cp.getElement(entry.getValue());
            if (transition == null) {
                throw new RuntimeException("[StateFactory] Incoming transition with ID " + entry.getValue() + " unknown!");
            }
            state.addInTransition(transition);
        }
        cp.stateInTransitionReferences.clear();

        for (Pair<Long, Long> entry : cp.stateOutTransitionReferences.getEntries()) {
            State state = (State) cp.getElement(entry.getKey());
            Transition transition = (Transition) cp.getElement(entry.getValue());
            if (transition == null) {
                throw new RuntimeException("[StateFactory] Outgoing transition with ID " + entry.getValue() + " unknown!");
            }
            state.addOutTransition(transition);
        }
        cp.stateOutTransitionReferences.clear();
    }
}
